src/model_repository_manager.h

```diff
@@ -316,33 +404,62 @@ class ModelRepositoryManager {
   /// \return True if the node is ready. False otherwise.
   bool CheckNode(
       DependencyNode* node,
-      const std::map<std::string, Status>& model_load_status);
-
-  Status CircularcyCheck(
-      DependencyNode* current_node, const DependencyNode* start_node);
+      const std::map<ModelIdentifier, Status>& model_load_status);
 
   bool ModelDirectoryOverride(
       const std::vector<const InferenceParameter*>& model_params);
 
-  std::set<std::string> repository_paths_;
   const bool autofill_;
   const bool polling_enabled_;
   const bool model_control_enabled_;
   const double min_compute_capability_;
 
   std::mutex poll_mu_;
-  ModelInfoMap infos_;
-
-  std::unordered_map<std::string, std::unique_ptr<DependencyNode>>
-      dependency_graph_;
-  std::unordered_map<std::string, std::unique_ptr<DependencyNode>>
-      missing_nodes_;
 
+  // Dependency graph
+
+  // WAR to avoid change in behavior (mainly error reporting) if model namespace
+  // is not enable.
+  std::function<Status(const std::string& model_name, ModelIdentifier* model_id)> find_identifier_fn_;
+  // A map from model name to model identifiers that share the same model name
+  std::unordered_map<std::string, std::set<ModelIdentifier>> global_map_;
+  DependencyGraph dependency_graph_;
+
+  // Repository specific..
+
+  const bool enable_model_namespacing_;
+  
+  // [FIXME] Better document below: 'infos_' is the in memory representation of
+  // repo polling results. It is a intermediate layer between the set of models
+  // in storage and the models being served, and it doesn't directly reflect the
+  // state of the two above.
+  // 
+  // There are three stages in terms of changing model:
+  //   - poll model
+  //   - resolve dependency
+  //   - load model
+  //
+  // 'infos_' is a long living object across model changes to serve as
+  // differentiator between polls (i.e. is the model newly added? modified?),
+  // so it is a subset of storage models at the time of polling. Note that
+  // it can be marked "dirty" to handle "fallback / revert" in model lifecycle:
+  // if the model is currently being served and re-load action is required, the
+  // lifecycle will keep the "older" model if the newer storage model is
+  // malformed => re-load failure. Marking info "dirty" in this case to ensure
+  // the model to be polled again even though the storage model hasn't been
+  // changed, for operation idempotence.
+  // (https://github.com/triton-inference-server/server/issues/3802)
+  ModelInfoMap infos_;
+  std::set<std::string> repository_paths_;
   // Mappings from (overridden) model names to a pair of their repository and
   // absolute path
+  // [FIXME] key should be updated to contain namespace as well, need to work with
+  // enable namespace. Need to revisit with repository side of things
```

Comment:
Basically when a mapping (i.e. `name_a -> name_b`) is added during repo registration `repo_1`, if we track it with model identifier instead of just the name, we can have repeated mapped name. For example, on top of `name_a -> name_b`, you register another repo (`repo_2`) that maps `name_c -> name_b`, currently it doesn't work because `name_b` is already a key, using model identifier will allow this because there are two keys `repo_1::name_b`, `repo_2::name_b`